agent/taskresource/logrouter/logrouter_unix.go

```diff
@@ -0,0 +1,355 @@
+func (logRouter *LogRouterResource) GetCreatedAt() time.Time {
+	logRouter.lock.RLock()
+	defer logRouter.lock.RUnlock()
+
+	return logRouter.createdAtUnsafe
+}
+
+// Create performs resource creation. This includes creating a config directory, a socket directory, and generating
+// a config file under the config directory.
+func (logRouter *LogRouterResource) Create() error {
+	// Fail fast if log router type is invalid.
+	if logRouter.logRouterType != LogRouterTypeFluentd && logRouter.logRouterType != LogRouterTypeFluentbit {
+		err := errors.New(fmt.Sprintf("invalid log router type: %s", logRouter.logRouterType))
+		logRouter.setTerminalReason(err.Error())
+		return err
+	}
+
+	err := logRouter.createDirectories()
+	if err != nil {
+		err = errors.Wrapf(err, "unable to initialize resource directory %s", logRouter.resourceDir)
+		logRouter.setTerminalReason(err.Error())
+		return err
+	}
+
+	err = logRouter.generateConfigFile()
+	if err != nil {
+		err = errors.Wrap(err, "unable to generate log router config file")
+		logRouter.setTerminalReason(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// createDirectories creates two directories:
+//  - $(DATA_DIR)/logrouter/$(TASK_ID)/config: used to store log router config file. The config file under this directory
+//    will be mounted to the log router container at an expected path.
+//  - $(DATA_DIR)/logrouter/$(TASK_ID)/socket: used to store the unix socket. This directory will be mounted to
+//    the log router container and it will generate a socket file under this directory. Containers that use log router
+//    will then use this socket to send logs to log router.
+func (logRouter *LogRouterResource) createDirectories() error {
+	configDir := filepath.Join(logRouter.resourceDir, "config")
+	err := logRouter.os.MkdirAll(configDir, os.ModePerm)
+	if err != nil {
+		return errors.Wrap(err, "unable to create config directory")
+	}
+
+	socketDir := filepath.Join(logRouter.resourceDir, "socket")
+	err = logRouter.os.MkdirAll(socketDir, os.ModePerm)
+	if err != nil {
+		return errors.Wrap(err, "unable to create socket directory")
+	}
+	return nil
+}
+
+// generateConfigFile generates a log router config file at $(RESOURCE_DIR)/config/fluent.conf.
+// This contains configs needed by the router container to properly route logs.
+func (logRouter *LogRouterResource) generateConfigFile() error {
+	config, err := logRouter.generateConfig()
+	if err != nil {
+		return errors.Wrap(err, "unable to generate log router config")
+	}
+
+	temp, err := logRouter.ioutil.TempFile(logRouter.resourceDir, tempFile)
+	if err != nil {
+		return err
+	}
+	defer temp.Close()
+	if logRouter.logRouterType == LogRouterTypeFluentd {
+		err = config.WriteFluentdConfig(temp)
+	} else {
+		err = config.WriteFluentBitConfig(temp)
+	}
+	if err != nil {
+		return err
+	}
+
+	err = temp.Chmod(os.FileMode(configFilePerm))
+	if err != nil {
+		return err
+	}
+
+	// Persist the config file to disk.
+	err = temp.Sync()
```

Comment:
I'm curious why you create a temp file, and then rename it to the desired path? What's the purpose of that?

Reply:
i think this ensures we never end up using a partially written config file (when write either hasn't finished or fails in the middle). Though we can trust resource dependency management to prevent that as well, i think there's no downside of having this to keep it safe. and it's the pattern being used to write agent state file and container metadata file anyway..

Reply:
we do this in other places of agent too https://github.com/aws/amazon-ecs-agent/pull/1765